ql/experimental/termstructures/crosscurrencyratehelpers.cpp

```diff
@@ -18,44 +18,154 @@
+                         Natural fixingDays,
+                         const Calendar& calendar,
+                         BusinessDayConvention convention,
+                         bool endOfMonth,
+                         const ext::shared_ptr<IborIndex>& idx) {
+            Schedule sch = legSchedule(evaluationDate, tenor, idx->tenor(), fixingDays, calendar,
+                                       convention, endOfMonth);
+            return IborLeg(sch, idx).withNotionals(1.0);
+        }
+
+        Real npvConstNotionalLeg(const Leg& iborLeg,
+                                 const Handle<YieldTermStructure>& discountCurveHandle) {
+            Date refDate = discountCurveHandle->referenceDate();
+            const YieldTermStructure& discountRef = **discountCurveHandle;
+            bool includeSettleDateFlows = true;
+            return CashFlows::npv(iborLeg, discountRef, includeSettleDateFlows, refDate) +
+                   discountRef.discount(iborLeg.back()->date()) - 1.0;
+        }
+
+        Real bpsConstNotionalLeg(const Leg& iborLeg,
+                                 const Handle<YieldTermStructure>& discountCurveHandle) {
+            const Spread basisPoint = 1.0e-4;
+            Date refDate = discountCurveHandle->referenceDate();
+            const YieldTermStructure& discountRef = **discountCurveHandle;
+            bool includeSettleDateFlows = true;
+            return CashFlows::bps(iborLeg, discountRef, includeSettleDateFlows, refDate) /
+                   basisPoint;
+        }
+
+        class ResettingLegHelper {
+          public:
+            explicit ResettingLegHelper(const YieldTermStructure& discountCurve,
+                                        const YieldTermStructure& foreignCurve)
+            : discountCurve_(discountCurve), foreignCurve_(foreignCurve) {}
+            Real discount(const Date& d) const {
+                return discountCurve_.discount(d);
+            }
+            Real adjustedNotional(const Date& d) const {
+                return foreignCurve_.discount(d) / discountCurve_.discount(d);
+            }
+
+          private:
+            const YieldTermStructure& discountCurve_;
+            const YieldTermStructure& foreignCurve_;
+        };
+
+        class ResettingLegNPVCalculator : public AcyclicVisitor, public Visitor<Coupon> {
+          public:
+            explicit ResettingLegNPVCalculator(const YieldTermStructure& discountCurve,
+                                               const YieldTermStructure& foreignCurve)
+            : helper_(discountCurve, foreignCurve), npv_(0.0) {}
+            void visit(Coupon& c) override {
+                Date start = c.accrualStartDate();
+                Date end = c.accrualEndDate();
+                Time accrual = c.accrualPeriod();
+                Real npv =
+                    helper_.adjustedNotional(start) *
+                    (-helper_.discount(start) + helper_.discount(end) * (1.0 + c.rate() * accrual));
```

Comment:
I'd try to separate the coupon and notional parts for clarity.  I'd also add a comment on how the notional part is estimated.

Reply:
I decomposed the pricing formula and added a clarifying comment.